### ISSUE TITLE
core (twisted from scratch) tutorial: web service example is flawed

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3735]
|<img alt="dhylton's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| dhylton reported|
|-|-|
|Trac ID|trac#3735|
|Type|defect|
|Created|2009-03-31 16:55:17Z|
|Branch|https://github.com/twisted/twisted/tree/mutate-users-3735|

in the ["Announce on Web, Too"](http://twistedmatrix.com/projects/core/documentation/howto/tutorial/protocol.html#auto4) section of the [protocol page](http://twistedmatrix.com/projects/core/documentation/howto/tutorial/protocol.html) of the [Twisted from Scratch tutorial](http://twistedmatrix.com/projects/core/documentation/howto/tutorial/index.html), FingerService creates a new user dictionary every time it re-reads the file (every 30 seconds) and new data is made available to finger requests - but the web resource is constructed with a direct reference to the first user dictionary and never sees any of the new user dictionaries.

file affected: [doc/core/howto/tutorial/listings/finger/finger15.tac#25837](doc/core/howto/tutorial/listings/finger/finger15.tac#25837)

_(thanks **exarkun**)_

<details><summary>Searchable metadata</summary>

```
trac-id__3735 3735
type__defect defect
reporter__dhylton dhylton
priority__normal normal
milestone__ 
branch__branches_mutate_users_3735 branches/mutate-users-3735
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1238518517000000 1238518517000000
changetime__1239836575000000 1239836575000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [26664]) Branching to 'mutate-users-3735'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

Fixed, I think.  No automated tests for this code, of course, which is too bad.  And manually testing is a serious pain in the ass, as the code wants to run as root, put files in `/etc` where I really don't want it to put them, etc.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph set owner to @exarkun|
|-|-|

Go ahead and merge this - it looks fine, as far as it goes.

Is there a ticket to make this code more test-friendly, and test it?  If not, file one, since it seems to keep coming up...

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [26698]) Propagate fix forward

refs [#3735](https://github.com/twisted/twisted/issues/3735)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun removed owner|
|-|-|

I overlooked something, please re-review.

[#10754](https://github.com/twisted/twisted/issues/10754) is a ticket for a TDD finger tutorial.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

With the caveat that nobody, especially nobody with my attention span, could review this level of (untested) duplication and be completely sure that it's all kosher, I've re-inspected this as best I can and the change appears to be properly propagated down through all subsequent versions of the tutorial.

Please land again.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set status to `closed`|
|-|-|

(In [26709]) Merge mutate-users-3735

Author: exarkun
Reviewer: glyph
Fixes: [#3735](https://github.com/twisted/twisted/issues/3735)

Fix the web-facing finger tutorial interface so that it notices updates to the
user file.  Previously user data was stored in a newly created dictionary once
in a while and the web server was never made aware of this new dictionary.  Now
the same dictionary is re-used each time user data is reloaded.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
